Devicemapper: Default to xfs instead of ext4

rhvgoyal opened the pull request:
Default to xfs instead of ext4. Also give a warning if user specified a filesystem using dm.fs option and base device already has a file system which is not same as dm.fs

icecrime commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -835,6 +843,18 @@ func getDeviceUUID(device string) (string, error) {
 	return uuid, nil
 }
 
+func getDeviceFS(device string) (string, error) {
```
Isn't that a duplicate of [`ProbeFsType`](https://github.com/docker/docker/blob/master/daemon/graphdriver/devmapper/mount.go#L41)?

rhvgoyal replied:
Right. Only difference seems to be that ProbeFsType() will return error if fs is anything other then ext4, xfs and btrfs. As we already use ProbeFsType(). So I can use ProbeFsType() too instead of using blkid.

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2258,6 +2291,37 @@ func (devices *DeviceSet) exportDeviceMetadata(hash string) (*deviceMetadata, er
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if strings.Contains(s.Text(), "xfs") {
```
this has to be `strings.HasSuffix(s.Text(), "\txfs")`

rhvgoyal replied:
Right. Otherwise selinuxfs will also be recognized as having valid xfs support. Will fix it.

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2258,6 +2291,37 @@ func (devices *DeviceSet) exportDeviceMetadata(hash string) (*deviceMetadata, er
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if strings.Contains(s.Text(), "xfs") {
+			if _, err := exec.LookPath("mkfs.xfs"); err == nil {
```
this has to be: 
```Go
_, err := exec.LookPath("mkfs.xfs")
return err == nil
```

rhvgoyal replied:
Or how about we simplify it a bit. We can first check if mkfs.xfs is present or not. If not, return error. If it is present, then we start looking at /proc/filesystems. That way instead of nesting the checks we do them one after the other. Makes reading code little easier.

tiborvass replied:
works for me

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2258,6 +2291,37 @@ func (devices *DeviceSet) exportDeviceMetadata(hash string) (*deviceMetadata, er
+
+	f, err := os.Open("/proc/filesystems")
+	if err != nil {
+		return false
```
`logrus.Warnf("Could not check if xfs is supported: %v", err)`

rhvgoyal replied:
Ok, will add.

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2258,6 +2291,37 @@ func (devices *DeviceSet) exportDeviceMetadata(hash string) (*deviceMetadata, er
+				return true
+			}
+		}
+	}
```
needs to check `s.Err()` and `logrus.Warnf` it if `!= nil`

rhvgoyal replied:
Ok, so at the end of s.Scan() for loop, I need to make sure scanner did not encounter any errors (except EOF) and if it did, then log a warning. Makes sense. Will do.

tiborvass replied:
no need to check for EOF, that's handled in the scanner's internals

tiborvass commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2269,7 +2333,7 @@ func NewDeviceSet(root string, doInit bool, options []string, uidMaps, gidMaps [
 		baseFsSize:            defaultBaseFsSize,
 		overrideUdevSyncCheck: defaultUdevSyncOverride,
-		filesystem:            "ext4",
+		defaultFilesystem:     determineDefaultFS(),
```
we shouldn't need to call this at creation, but only if user did not specify `dm.fs`, so after the options are parsed and checked if `devices.userFilesystem` is not set

rhvgoyal replied:
ok, will do.

rhvgoyal replied:
In fact I think I can get rid of devices.defaultFilesystem. As devices.defaultFilesystem is consulted only at one place (during file system creation), I think I will not even store it in devices. Just save it in a local variable and discard it after that.

tiborvass replied:
whatever works. The way I'll test it is by upgrading docker with this patch and see if my images are still there and runnable.

rhvgoyal committed:
devmapper: Switch to xfs as default filesystem if supported

If platform supports xfs filesystem then use xfs as default filesystem 
for container rootfs instead of ext4. Reason being that ext4 is pre-allcating
lot of metadata (around 1.8GB on 100G thin volume) and that can take long
enough on AWS storage that systemd times out and docker fails to start.

If one disables pre-allocation of ext4 metadata, then it will be allocated
when containers are mounted and we will have multiple copies of metadata
per container. For a 100G thin device, it was around 1.5GB of metadata
per container.

ext4 has an optimization to skip zeroing if discards are issued and
underlying device guarantees that zero will be returned when discarded
blocks are read back. devicemapper thin devices don't offer that guarantee
so ext4 optimization does not kick in. In fact given discards are optional
and can be dropped on the floor if need be, it looks like it might not be
possible to guarantee that all the blocks got discarded and if read back
zero will be returned.

Signed-off-by: Anusha Ragunathan <anusha@docker.com>
Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
devmapper: Warn if user specified a filesytem and base device already has fs

If user wants to use a filesystem it can be specified using dm.fs=<filesystem>
option. It is possible that docker already had base image and a filesystem
on that. Later if user wants to change file system using dm.fs= option
and restarts docker, that's not possible. Warn user about it.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@tiborvass  Pushed V2 of patches. I think I have taken care of all review comments. PTAL.

crosbymichael commented:
LGTM

anusha-ragunathan commented:
LGTM

tiborvass commented:
LGTM

bmangold commented:
I ran engine 1.9.0 + this change on aws for rhel 7.1 SELinux and centos ... looks good.  No issues with starting the daemon after install.
